keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,286 @@
+
+To properly support RayCluster, Kueue needs to gracefully handle the scale up and scale down of RayCluster nodes. Concretely this means that scaling the resources used by a RayCluster is appropriately reflected in the respective ClusterQueue without needing to suspend the entire cluster.
+
+### Goals
+
+- Gracefully handle resize operations for Kueue jobs (i.e. update quota usage without suspend, enqueue scale ups)
+- Autoscaling RayCluster works with Kueue (MVP)
+
+### Non-Goals
+
+- Vertical scaling of workloads – Kueue will only handle resize operations that scale Pods horizontally 
+- Support resize for other Kueue jobs such as QueuedWorkload, JobSet, etc (future)
+- Resizing of RayJobs
+- Partial Preemption
+
+## Proposal
+
+Update the Job framework reconciler and introduce new controllers to orchestrate dynamic resizing of jobs. We are only interested in horizontal scaling of jobs (e.g. scaling more replicas). At a high level, this will be accomplished by:
+- Creating Workload Slice objects that represent incremental scale up of jobs. This gives us per-replica control of workload admission. Workload Slices will be garbage collected and consolidated with their parent workloads after successful admission.
+- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
+- Dynamically adjust quotas in ClusterQueue based on scaling events.
+
+For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.
+
+### User Stories
+
+#### Story 1 - RayCluster w/ autoscaling
+
+1. The user creates a RayCluster.
+2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
+3. User updates RayCluster to enable autoscaling
+4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 
+
+## Design Details
+
+### Workload Slices
+To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.
+
+The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.
+
+
+### Creating Workload Slices
+
+The [GenericJob interface](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go#L30-L55) will be updated to handle resize operations of jobs.
+
+```golang
+type GenericJob interface {
+  ...
+  ...
+  ResizeJob(wl *kueue.Workload) error
+}
+```
+Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 
```

Comment:
If the implementation of `ResizeJob` is optional, then it should be part of a new interface.

Reply:
Something like https://github.com/kubernetes-sigs/kueue/blob/b6f7a1c9cfd58fbc25c0909d7910fddb1eba7e8f/pkg/controller/jobframework/interface.go#L60 ?